streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java

```diff
@@ -307,6 +307,25 @@
     public static final String APPLICATION_SERVER_CONFIG = "application.server";
     private static final String APPLICATION_SERVER_DOC = "A host:port pair pointing to a user-defined endpoint that can be used for state store discovery and interactive queries on this KafkaStreams instance.";
 
+    /** {@code assignment.acceptable.recovery.lag} */
+    public static final String ASSIGNMENT_ACCEPTABLE_RECOVERY_LAG_CONFIG = "assignment.acceptable.recovery.lag";
+    private static final String ASSIGNMENT_ACCEPTABLE_RECOVERY_LAG_DOC = "The maximum acceptable lag (number of offsets to catch up) for an active task to be assignable to a client. " +
+                                                                    "Should correspond to a recovery time of well under a minute for a given workload. Must be at least 0.";
+
+    /** {@code assignment.balance.factor} */
+    public static final String ASSIGNMENT_BALANCE_FACTOR_CONFIG = "assignment.balance.factor";
+    private static final String ASSIGNMENT_BALANCE_FACTOR_DOC = "Target maximum difference in the number of active tasks assigned to the thread with the most tasks and the thread with the least in a steady-state assignment.";
+
+    /** {@code assignment.max.extra.replicas} */
+    public static final String ASSIGNMENT_MAX_EXTRA_REPLICAS_CONFIG = "assignment.max.extra.replicas";
+    private static final String ASSIGNMENT_MAX_EXTRA_REPLICAS_DOC = "The maximum number of extra replicas that can be assigned at once for the purpose of keeping the task available on one instance while it is warming up on " +
```

Comment:
I've been going back and forth on whether to add this as a "per-cluster" or "per-instance" config, @vvcephei  and I discussed briefly but @cadonna any opinion? If we do go with cluster-wide I think we should make the default pretty big, but I'm now leaning towards towards making it per-instance. Thoughts?

Reply:
Also pretty significantly modified the original name, I didn't feel "max.task.migrations" got to the point and was imagining a lot of users asking us about this. Not sure "max.extra.replicas" is the right way to go either, any other suggestions (or arguments in favor of "max.task.migrations")?

Reply:
> whatever is specified on the group leader is the truth
Yes, and this applies to all the rebalance/assignment related configs.

For the per-instance option, I wasn't thinking we'd let each client send a different value (although we certainly could do this), just that we'd take whatever value the leader has and apply it to all instances. This way it doesn't complicate the algorithm, and we don't have to add it to the subscription: and my impression is that surely in practice users don't spin up instances with drastically different performance characteristics? But, if we go with the per-instance config and say only the leader's value will be chosen, and it turns out users would like to be able to set it differently on different instances, we can easily bump the protocol version in the next release and tell users they can now set it differently for different instances.

I think a per-thread max may be a bit weird for users to reason about, but let's first decide if we do/don't want to make it cluster-wide before going on to instance vs. thread. One of the main arguments I still see for making it instance-wide, is that this is supposed to be a throttling mechanism and the most natural way to think of it in that light is probably on a cluster-wide basis, ie, "don't let this app ever use more than 4 extra tasks of resources".

This of course limits how many instances can be brought up at once while maintaining high availability; on the other hand, if you really are resource constrained (and presumably you wouldn't set this to a low value otherwise) you can't bring up a large number of instances at the same time anyway, without hitting your system/app limits

Reply:
Now I'm leaning towards making it cluster-wide, but setting it to a very high default and warning users that lowering it creates a tradeoff between high availability and time to scale out

Reply:
I took again a look at the KIP and I agree with you that the main motivation is throttling the usage of cluster resources not the instance resources. If this is still true then cluster-wide makes more sense to me.

Reply:
Sounds good -- do you think 2 is a reasonable default for it as a cluster-wide config? I think that's way too low, I'm actually inclined to set it to `MAX_VALUE` and maybe bump it to "Medium" priority so users consider whether to lower it. WDYT @cadonna @vvcephei ?

Reply:
I would rather be conservative here, since too many warm-up replicas might cost more money in a cloud setting. Wouldn't them?
Bump the config to medium priority seems reasonable to me in any case.

Reply:
That's true...but a lower number will essentially make high availability an "opt-in" feature, which it's not (and should not be). Also, how do we choose a reasonable number? It doesn't make sense to optimize for a specific case without knowing what the specific restraints. If we set it to "5", then how many users will still need to change it because their cluster can actually tolerate 10, or only 1, or 50 (+ all the users who want high availability no matter what). If we set it to `MAX_VLUE` then all those same users will still have to set it, minus maybe a handful for whom we happened to guess right that "5" was the allowable number.
It seems better to not set the limit and document very clearly that users with resource constraints should set a limit, than guess wildly at what that limit should be while also restricting the high availability for all by default

Reply:
Thanks for the discussion, all.

One thing to note is that this config has "no" impact on high availability. Anyone who wants HA would configure num.standby.replicas, and this config applies on top of that one (for this reason, I like the name you picked. It is "extra".)

I think you can make a case for per-node or per-cluster, but if we assume we only want to add one config, I think per-cluster is more valuable, since it lets you protect the brokers from overload, which a per-node config may not.

Regarding whether we set the default limit low or high, I'd advocate for low. Some clusters are undoubtedly running close to the limits of their disk space or broker capacity, so suddenly letting every node double its traffic would result in serious operational consequences. On the other hand, if we start out low, then the probability of a crash becomes much lower, and the only problem is that the overall balancing process takes a long time. But, since the config is set low, there's a low impact on processing capacity while it's happening, so maybe there's no real impact over that long time.

If you buy the argument that a low default is good, then the obvious choice is "1", but that would take a _really_ long time to complete balancing. The default of "2" is basically a compromise. It lets you balance twice as fast, and it's "probably" still low enough to cause a problem for no one. "5" also seems fine-ish, but the farther from "1" you move, the riskier the choice is.

One final thought, when you say that some people would "need to change it". It seems like this population is restricted to the people who really need to make sure that balancing happens absolutely as fast as possible (maybe because their application can't keep up unless the whole cluster is processing, or because they're scaling up to cope with a spike in input data). Hopefully, these people know that they need extra horsepower and are motivated to read the docs and locate the config they need to change. Also, hopefullly, this isn't the common case.

Reply:
Yep, thanks for the clarification, I was misinterpreting how this config would be applied. FWIW I stand by the claim "2 is 2 low" :

 you kind on touch on this in the 4th paragraph ("Regarding..."), but I'm just wondering what is likely to be the state of the cluster at the moment of scaling out. In other words, what is the bottleneck driving the need to bring up more instances? It can't be broker traffic, because bringing up another instance will _not_ help with that. More likely it's something like disk, or consumer lag.

If the bottleneck is disk, throttling the number of warmup replicas will have absolutely zero impact on the immediate state as the original owner will remain the owner until the new instance is ready. Throttling will only serve to slow this down, and increase the likelihood we run out of disk before migrating enough tasks away*

If it's consumer lag, then again we just delay the ability to catch up as the over-constrained instance will have to continue processing active tasks for longer while waiting for a warmup replica to be created.

So, I agree now that this has no implications for high availability, but it _does_ seem to have implications for "smooth scaling out", which is the literal title of this KIP 😉 . Of course, in the real world you may have two bottlenecks/constraints at the same time, so just because you're scaling out doesn't mean you don't also have to worry about broker traffic. But I would argue that that's something the user should customize if they expect it to be a constraint, and we shouldn't limit the cluster's ability to quickly relax other bottlenecks which, after all, is most likely the reason for scaling out in the first place.

But, since I have less (ie, no) experience on the operational side of things I'm happy to trust your intuition. Sorry to keep dragging this out, I'll shut up now

Reply:
Haha, no problem. I think your reasoning makes sense. It's certainly true that if the cluster is at a moment when it direly needs scale-out to happen as fast as possible, then the operator would likely need to un-throttle the task migration. As you alluded to in your comment, I have myself been in this exact situation. 

It's tricky, though, because you can easily turn one incident into a much worse incident with this kind of emergency activity. For example, you app is not keeping up, so you double its size, and all the extra migration load now overwhelms the broker, causing a company-wide outage. Not saying this would happen, but it's the basic kind of thought that's behind my desire to err on the side of "slow and steady" with the default.

At least, if that _did_ happen, it would be the operator's fault for de-throttling too much, not ours for selecting a dangerous default. I'm not just passing the buck with this statement. The operator is right there, in the moment. When the broker craps out, they know exactly what they've done and are in a good position to react immediately. If the fault were our bad default, then it might be much longer before anyone even reads the docs and learns that there _is_ a throttling config, and that they can set it lower.

Not sure if this really adds much to the conversation, since you already yielded the point, but it felt like a little more elaboration might be nice.

Reply:
Fair enough. Just wanted to put it all out there -- good chat 😄